viz_scripts/survey_responses.ipynb

```diff
@@ -0,0 +1,343 @@
+    "    \n",
+    "    for survey_name in sheet_list:\n",
+    "        url = sheet_list[survey_name]\n",
+    "        result = urllib.request.urlopen(url)\n",
+    "        doc = minidom.parse(result) \n",
+    "        \n",
+    "        #nested dictionaries to keep surveys grouped\n",
+    "        survey_questions = {}\n",
+    "\n",
+    "        labels = doc.getElementsByTagName(\"label\") \n",
+    "        for label in labels:\n",
+    "            if(bool(label.parentNode.getAttribute(\"ref\"))):\n",
+    "                #label appearance = not a question?\n",
+    "                #nodeName input -- to word cloud as a later addition?\n",
+    "                if label.parentNode.getAttribute(\"appearance\") != \"label\" and label.parentNode.nodeName != \"input\": \n",
+    "                    survey_questions[str(label.parentNode.getAttribute(\"ref\").split('/')[-1])] = label.firstChild.data\n",
+    "            \n",
+    "            if label.parentNode.nodeName == 'item':\n",
+    "                if label.parentNode.parentNode.getAttribute(\"appearance\") == \"likert\":\n",
+    "                    if label.parentNode.getElementsByTagName(\"value\")[0].firstChild.data not in opt_dict.keys():\n",
+    "                        opt_dict[label.parentNode.getElementsByTagName(\"value\")[0].firstChild.data] = label.parentNode.getElementsByTagName(\"value\")[0].firstChild.data\n",
+    "                elif label.parentNode.getElementsByTagName(\"value\"):\n",
+    "                    if label.parentNode.getElementsByTagName(\"value\")[0].firstChild.data not in opt_dict.keys():\n",
+    "                        opt_dict[label.parentNode.getElementsByTagName(\"value\")[0].firstChild.data] = label.firstChild.data  \n",
+    "                elif label.parentNode.getElementsByTagName(\"name\"):\n",
+    "                    if label.parentNode.getElementsByTagName(\"name\")[0].firstChild.data not in opt_dict.keys():\n",
+    "                        opt_dict[label.parentNode.getElementsByTagName(\"name\")[0].firstChild.data] = label.firstChild.data  \n",
+    "        \n",
+    "        quest_dict[survey_name] = survey_questions\n",
+    "        \n",
+    "    return opt_dict, quest_dict\n",
+    "\n",
+    "\n",
+    "#input: dataframe containing all trips that have non-blank user_input\n",
+    "#output: dataframe with questions in the columns and answers in the rows\n",
+    "#for loop will run n survey responses times (this could get big!)\n",
+    "def create_dataframe(df_trips_w_surveys):\n",
+    "    df = df_trips_w_surveys.reset_index()\n",
+    "    rows = []\n",
+    "    for i in range(len(df)):\n",
+    "        data_key = list(df.loc[i].user_input['trip_user_input']['data']['jsonDocResponse'].keys())[0]\n",
+    "        row = pd.json_normalize(df.loc[i].user_input['trip_user_input']['data']['jsonDocResponse'][data_key])\n",
+    "        row['user_id'] = df.loc[i].user_id\n",
+    "        rows.append(row)\n",
```

Comment:
why are we json_normalizing row by row?
I think this is because we could have multiple responses to the survey for a trip and we want to use the most recent one? If so, the pythonic option would be something like

`pd.json_normalize(df.apply(pick_latest_response))` where `pick_latest_response` is a lambda that selects the most recent answer. In general, you want to use vectorized operations instead of iterating over for loops for greater efficiency.

Reply:
addressed in [548a449](https://github.com/e-mission/em-public-dashboard/pull/124/commits/548a449d6b46130a0441eb39b372adca0ee56459)